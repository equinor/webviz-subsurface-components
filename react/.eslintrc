{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "extends":  [
    "plugin:react/recommended",
    "eslint:recommended",
    "plugin:prettier/recommended"
  ],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx"],
      "plugins": ["@typescript-eslint"],
      "extends": [
        "plugin:react/recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:prettier/recommended"
      ]
    }
  ],
  "settings": {
    "react": {
      "version": "detect"
    }
  },
  "env": {
    "amd": true,
    "browser": true,
    "es6": true,
    "jasmine": true,
    "node": true
  }
}
