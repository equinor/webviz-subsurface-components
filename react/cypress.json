{
    "baseUrl": "http://localhost:6006/?path=/story/deckglmap--default",
    "video": true
}
