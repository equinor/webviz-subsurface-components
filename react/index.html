<html>
    <head>
        <title>Demo application of webviz-subsurface-components</title>
    </head>
    <body>
        <div id='root'></div>
        <link rel="stylesheet" type="text/css" href="output.css">
        <script src="output.js"></script>
    </body>
</html>
