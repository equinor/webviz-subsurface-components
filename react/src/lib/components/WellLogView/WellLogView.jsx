import React, { Component } from 'react';
import {
  LogViewer, UIHelper,
}  from '@equinor/videx-wellog';

import './styles.scss';

import { select } from 'd3';

import createTracks from './tracks';

function addRubberbandOverlay(instance) {
  const rubberBandSize = 9;
  const offset = (rubberBandSize - 1) / 2;
  const rbelm = instance.overlay.create('rubber-band', {
    onMouseMove: event => {
      const { y } = event;
      event.target.style.top = `${y - (offset + 0.5)}px`;
      event.target.style.visibility = 'visible';
    },
    onMouseExit: event => {
      event.target.style.visibility = 'hidden';
      if (instance.options.rubberbandExit) {
        instance.options.rubberbandExit({
          source: instance,
        });
      }
    }
  });

  const rb = select(rbelm).classed('rubber-band', true)
    .style('height', `${rubberBandSize}px`)
    .style('background-color', 'rgba(255,0,0,0.1)')
    .style('visibility', 'hidden');

  rb.append('div')
    .style('height', '1px')
    .style('background-color', 'rgba(255,0,0,0.7)')
    .style('position', 'relative')
    .style('top', `${offset}px`);
}

function addReadoutOverlay(instance) {
  const elm = instance.overlay.create('depth', {
    onClick: event => {
      const {
        target,
        caller,
        y,
      } = event;
      const md = caller.scale.invert(y);
      target.textContent = Number.isFinite(md)
        ? `Pinned MD: ${md.toFixed(1)}`
        : '-';
      target.style.visibility = 'visible';
    },
    onMouseMove: event => {
      const {
        target,
        caller,
        y,
      } = event;
      const md = caller.scale.invert(y);
      target.textContent = Number.isFinite(md)
        ? `MD: ${md.toFixed(1)}`
        : '-';
      target.style.visibility = 'visible';
    },
    onMouseExit: event => {
      event.target.style.visibility = 'hidden';
    },
    onRescale: event => {
      event.target.style.visibility = 'visible';
      event.target.textContent = `Zoom: x${event.transform.k.toFixed(1)}`;
    }
  });
  elm.style.visibility = 'hidden';
  elm.style.display = 'inline-block';
  elm.style.padding = '2px';
  elm.style.borderRadius = '4px';
  elm.style.textAlign = 'right';
  elm.style.position = 'absolute';
  elm.style.backgroundColor = 'rgba(0,0,0,0.5)';
  elm.style.color = 'white';
  elm.style.right = '5px';
  elm.style.bottom = '5px';
}

function addPinnedValueOverlay(instance) {
  const rubberBandSize = 9;
  const offset = (rubberBandSize - 1) / 2;
  const rbelm = instance.overlay.create('pinned', {
    onClick: event => {
      const { y } = event;
      event.target.style.top = `${y - (offset + 0.5)}px`;
      event.target.style.visibility = 'visible';
    },
    onMouseExit: event => {
      event.target.style.visibility = 'hidden';
    }
  });

  const rb = select(rbelm).classed('pinned', true)
    .style('height', `${rubberBandSize}px`)
    .style('background-color', 'rgba(0,0,0,0.1)')
    .style('position', 'absolute')
    .style('width', '100%')
    .style('visibility', 'hidden');

  rb.append('div')
    .style('height', '1px')
    .style('background-color', 'rgba(0,255,0,0.7)')
    .style('position', 'relative')
    .style('top', `${offset}px`);
}




class WellLogView extends Component {
  componentDidMount() {
    const logController = new LogViewer({
      showLegend: true,
      horizontal: false,
    });
    const tracks = createTracks(true);
this.container.style.height = "96%"
    logController.init(this.container).setTracks(tracks);
this.container.classList.add('wellog');

    addReadoutOverlay(logController);
    addRubberbandOverlay(logController);
    addPinnedValueOverlay(logController);
  }

  render() {
    return (
      <div>
        <div ref={el => { this.container = el; }} />
      </div>
    );
  }
}

export default WellLogView
