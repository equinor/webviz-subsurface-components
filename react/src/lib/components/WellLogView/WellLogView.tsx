import React, { Component } from 'react';
import {
    LogViewer, InterpolatedScaleHandler,
    UIHelper,
}  from '@equinor/videx-wellog';

import './styles.scss';

import { select } from 'd3';

import createTracks from './tracks';

function addRubberbandOverlay(instance) {
  const rubberBandSize = 9;
  const offset = (rubberBandSize - 1) / 2;
  const rbelm = instance.overlay.create('rubber-band', {
    onMouseMove: event => {
      const { y } = event;
      event.target.style.top = `${y - (offset + 0.5)}px`;
      event.target.style.visibility = 'visible';
    },
    onMouseExit: event => {
      event.target.style.visibility = 'hidden';
      if (instance.options.rubberbandExit) {
        instance.options.rubberbandExit({
          source: instance,
        });
      }
    }
  });

  const rb = select(rbelm).classed('rubber-band', true)
    .style('height', `${rubberBandSize}px`)
    .style('background-color', 'rgba(255,0,0,0.1)')
    .style('visibility', 'hidden');

  rb.append('div')
    .style('height', '1px')
    .style('background-color', 'rgba(255,0,0,0.7)')
    .style('position', 'relative')
    .style('top', `${offset}px`);
}

function addReadoutOverlay(instance) {
  const elm = instance.overlay.create('depth', {
    onClick: event => {
      const {
        target,
        caller,
        y,
      } = event;
      const md = caller.scale.invert(y);
      target.textContent = Number.isFinite(md)
        ? `Pinned MD: ${md.toFixed(1)}`
        : '-';
      target.style.visibility = 'visible';
    },
    onMouseMove: event => {
      const {
        target,
        caller,
        y,
      } = event;
      const md = caller.scale.invert(y);
      target.textContent = Number.isFinite(md)
        ? `MD: ${md.toFixed(1)}`
        : '-';
      target.style.visibility = 'visible';
    },
    onMouseExit: event => {
      event.target.style.visibility = 'hidden';
    },
    onRescale: event => {
      event.target.style.visibility = 'visible';
      event.target.textContent = `Zoom: x${event.transform.k.toFixed(1)}`;
    }
  });
  elm.style.visibility = 'hidden';
  elm.style.display = 'inline-block';
  elm.style.padding = '2px';
  elm.style.borderRadius = '4px';
  elm.style.textAlign = 'right';
  elm.style.position = 'absolute';
  elm.style.backgroundColor = 'rgba(0,0,0,0.5)';
  elm.style.color = 'white';
  elm.style.right = '5px';
  elm.style.bottom = '5px';
}

function addPinnedValueOverlay(instance) {
  const rubberBandSize = 9;
  const offset = (rubberBandSize - 1) / 2;
  const rbelm = instance.overlay.create('pinned', {
    onClick: event => {
      const { y } = event;
      event.target.style.top = `${y - (offset + 0.5)}px`;
      event.target.style.visibility = 'visible';
    },
    onMouseExit: event => {
      event.target.style.visibility = 'hidden';
    }
  });

  const rb = select(rbelm).classed('pinned', true)
    .style('height', `${rubberBandSize}px`)
    .style('background-color', 'rgba(0,0,0,0.1)')
    .style('position', 'absolute')
    .style('width', '100%')
    .style('visibility', 'hidden');

  rb.append('div')
    .style('height', '1px')
    .style('background-color', 'rgba(0,255,0,0.7)')
    .style('position', 'relative')
    .style('top', `${offset}px`);
}

interface Props {
    data: any
};

class WellLogView extends Component<Props> {
    container?: HTMLElement;
    logController?: LogViewer;

    constructor(props) {
        super(props)
        //alert("props=" + props)

        this.container = undefined
        this.logController = undefined

        //this.setTracks()
    }

    componentDidMount() {
        this.createLogViewer()
    }

    componentDidUpdate(prevProps) {
        // Typical usage (don't forget to compare props):
        if (this.props.data !== prevProps.data) {
            //setState(???)
            //this.setTracks();
            this.createLogViewer() // recreate LogViewer
        }
    }

    createLogViewer() {
        let axes = {
            primary: "md",
            secondary: "tvd"
        }
        const {
            tracks,
            minmaxPrimaryAxis,
            interpolator,
        } = createTracks(this.props.data, axes);

        let scaleHandler = new InterpolatedScaleHandler(interpolator);

        if (this.logController) { // remove old LogViewer
            this.logController.onUnmount(); // 
            this.logController = undefined;
        }
        if (this.container) {
            this.container.innerHTML = ""; // delete all old LogViewer elements

            // create new LogViewer
            this.logController = new LogViewer({
                showLegend: true,
                horizontal: false,
                domain: minmaxPrimaryAxis,
                scaleHandler: scaleHandler
            });

            this.logController.init(this.container);

            addReadoutOverlay(this.logController);
            addRubberbandOverlay(this.logController);
            addPinnedValueOverlay(this.logController);

            //this.setTracks();
            this.logController.setTracks(tracks);
        }
    }

    setTracks() {
        if (!this.logController)
            return;
        let axes = {
            primary: "md",
            secondary: "tvd"
        }

        const {
            tracks,
            minmaxPrimaryAxis,
            interpolator
        } = createTracks(this.props.data, axes);
        //alert(minmaxPrimaryAxis)

        //let scaleHandler = new InterpolatedScaleHandler(interpolator);
        //this.logController.scaleHandler(scaleHandler);// not worked??

        this.logController.domain = minmaxPrimaryAxis;
        this.logController.setTracks(tracks);
    }

    render() {
        return (
            <div>
                <table height='100%' width='100%'>
                    <tr>
                        <td className='wellog' ref={el => { this.container = el as HTMLElement; }} />
                        {/*
                        <td valign='top'>
                            <fieldset>
                                <legend>Scale tracks</legend>
                                <input type='checkbox' checked={true} onChange={function () { return false; }} /> MD<br />
                                <input type='checkbox' /> TVD
                            </fieldset>
                            <fieldset>
                                <legend>Readout</legend>
                                <table width='100' id='well-log-readout'>
                                    <tr><td>MD</td><td>mtr</td><td>2345.6</td></tr>
                                    <tr><td>TVD</td><td>mtr</td><td>1234.5</td></tr>
                                    <tr><td>HKLA</td><td>ohm.m</td><td>110.5</td></tr>
                                    <tr><td>HKLX</td><td></td><td>80.15</td></tr>
                                    <tr><td>...</td></tr>
                                </table>
                            </fieldset>
                        </td>
                        */}
                    </tr>
                </table>
            </div>
        );
    }
}

export default WellLogView
