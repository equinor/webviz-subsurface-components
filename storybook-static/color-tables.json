[
    {
        "name": "Physics",
        "discrete": false,
        "description": "Full options color table",
        "colorNaN": [255, 255, 255],
        "colorBelow": [255, 0.0, 0.0],
        "colorAbove": [0.0, 0.0, 255],
        "colors": [
            [0.0, 255, 0.0, 0.0],
            [0.25, 182, 182, 0.0],
            [0.5, 0.0, 255, 0.0],
            [0.75, 0.0, 182, 182],
            [1.0, 0.0, 0.0, 255]
        ]
    },
    {
        "name": "Physics reverse",
        "discrete": false,
        "colors": [
            [0.0, 0.0, 0.0, 255],
            [0.25, 0.0, 182, 182],
            [0.5, 0.0, 255, 0.0],
            [0.75, 182, 182, 0.0],
            [1.0, 255, 0.0, 0.0]
        ]
    },
    {
        "name": "Rainbow",
        "discrete": false,
        "colors": [
            [0.0, 255, 0.0, 0.0],
            [0.2, 182, 182, 0.0],
            [0.4, 0.0, 255, 0.0],
            [0.6, 0.0, 182, 182],
            [0.8, 0.0, 0.0, 255],
            [1.0, 182, 0.0, 182]
        ]
    },
    {
        "name": "Rainbow reverse",
        "discrete": false,
        "colors": [
            [0.0, 182, 0.0, 182],
            [0.2, 0.0, 0.0, 255],
            [0.4, 0.0, 182, 182],
            [0.6, 0.0, 255, 0.0],
            [0.8, 182, 182, 0.0],
            [1.0, 255, 0.0, 0.0]
        ]
    },
    {
        "name": "Colors_set_1",
        "discrete": "true",
        "colors": [
            [0, 255, 13, 186],
            [1, 255, 64, 53],
            [2, 247, 255, 164],
            [3, 112, 255, 97],
            [4, 9, 254, 133],
            [5, 254, 4, 135],
            [6, 255, 5, 94],
            [7, 32, 50, 255],
            [8, 109, 255, 32],
            [9, 254, 146, 92],
            [10, 185, 116, 255],
            [11, 255, 144, 1],
            [12, 157, 32, 255],
            [13, 255, 26, 202],
            [14, 73, 255, 35]
        ]
    },
    {
        "name": "Stratigraphy",
        "discrete": true,
        "colorNaN" : [255, 64, 64],
        "colors": [
            [0, 255, 193, 0],
            [1, 255, 120, 61],
            [2, 255, 155, 76],
            [3, 255, 223, 161],
            [4, 226, 44, 118],
            [5, 255, 243, 53],
            [6, 255, 212, 179],
            [7, 255, 155, 23],
            [8, 255, 246, 117],
            [9, 255, 241, 0],
            [10, 255, 211, 178],
            [11, 255, 173, 128],
            [12, 248, 152, 0],
            [13, 154, 89, 24],
            [14, 0, 138, 185],
            [15, 82, 161, 40],
            [16, 219, 228, 163],
            [17, 0, 119, 64],
            [18, 0, 110, 172],
            [19, 116, 190, 230],
            [20, 0, 155, 212],
            [21, 0, 117, 190],
            [22, 143, 40, 112],
            [23, 220, 153, 190],
            [24, 226, 44, 118],
            [25, 126, 40, 111],
            [26, 73, 69, 43],
            [27, 203, 63, 42],
            [28, 255, 198, 190],
            [29, 135, 49, 45],
            [30, 150, 136, 120],
            [31, 198, 182, 175],
            [32, 166, 154, 145],
            [33, 191, 88, 22],
            [34, 255, 212, 179],
            [35, 251, 139, 105],
            [36, 154, 89, 24],
            [37, 186, 222, 200],
            [38, 0, 124, 140],
            [39, 87, 84, 83]
        ]
    },
    {
        "name": "Colors_set_3",
        "discrete": "true",
        "colors": [
            [0, 120, 181, 255],
            [1, 255, 29, 102],
            [2, 247, 255, 173],
            [3, 239, 157, 255],
            [4, 186, 255, 236],
            [5, 46, 255, 121],
            [6, 212, 255, 144],
            [7, 165, 255, 143],
            [8, 122, 255, 89],
            [9, 255, 212, 213]
        ]
    },
    {
        "name": "Porosity",
        "discrete": false,
        "colors": [
            [0.00, 255, 246, 117],
            [0.11, 255, 243, 53],
            [0.18, 255, 241, 0],
            [0.25, 155, 193, 0],
            [0.32, 255, 155, 23],
            [0.39, 255, 162, 61],
            [0.46, 255, 126, 45],
            [0.53, 227, 112, 24],
            [0.60, 246, 96, 31],
            [0.67, 229, 39, 48],
            [0.74, 252, 177, 170],
            [0.81, 236, 103, 146],
            [0.88, 226, 44, 118],
            [1.00, 126, 40, 111]
        ]
    }
]
