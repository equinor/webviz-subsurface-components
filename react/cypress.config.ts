import { defineConfig } from "cypress";
import getCompareSnapshotsPlugin from "cypress-image-diff-js/dist/plugin";

export default defineConfig({
  component: {
    devServer: {
      framework: "create-react-app",
      bundler: "webpack",
    },
    video:false,
    setupNodeEvents(on, config) {
      getCompareSnapshotsPlugin(on, config),
      on('before:browser:launch', (browser = {}, launchOptions) => {
        if (browser.family === 'chromium' && browser.name !== 'electron') {
          launchOptions.args.push('--start-fullscreen')
      
          return launchOptions
        }
        if (browser.name === 'electron') {
          launchOptions.preferences.fullscreen = true
      
          return launchOptions
        }
      })
    }
  },
});
