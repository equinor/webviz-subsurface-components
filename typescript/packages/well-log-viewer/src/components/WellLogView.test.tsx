/* eslint-disable @typescript-eslint/no-var-requires */
import { colorTables } from "@emerson-eps/color-tables";
import { render } from "@testing-library/react";
import "jest-styled-components";
import React from "react";
import WellLogView from "./WellLogView";

const exampleTemplate = require("../../../../../example-data/welllog_template_1.json");
const exampleWellLog = {
    header: {},
    curves: [],
    data: [],
};
const exampleColorTable = colorTables;

window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
        disconnect: jest.fn(),
        observe: jest.fn(),
        unobserve: jest.fn(),
    }));

describe("Test Well Log View Component", () => {
    it("snapshot test", () => {
        const { container } = render(
            <WellLogView
                welllog={exampleWellLog}
                options={{
                    checkDatafileSchema: true,
                    hideTrackLegend: true,
                    hideTrackTitle: true,
                }}
                template={exampleTemplate}
                colorTables={exampleColorTable}
                primaryAxis={"md"}
                axisTitles={{
                    md: "MD",
                    tvd: "TVD",
                    time: "TIME",
                }}
                axisMnemos={{
                    md: ["DEPTH", "DEPT", "MD", "TDEP", "MD_RKB"],
                    tvd: ["TVD", "TVDSS", "DVER", "TVD_MSL"],
                    time: ["TIME"],
                }}
            />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
