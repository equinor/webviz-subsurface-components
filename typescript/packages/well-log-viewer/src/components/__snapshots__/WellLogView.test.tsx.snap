// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Well Log View Component snapshot test 1`] = `
<div
  class="welllogview"
  style="flex-direction: column;"
>
  <div
    class="view"
  >
    <div
      class="container log-controller"
    >
      <div
        class="overlay"
        style="width: 0px; left: 0px; overflow: hidden;"
      >
        <div
          class="depth"
          style="position: absolute; pointer-events: none; visibility: hidden;"
        />
        <div
          class="rubber-band"
          style="position: relative; pointer-events: none; height: 9px; width: 100%; visibility: hidden;"
        >
          <div
            style="height: 1px; width: 100%; top: 4.5px; position: relative;"
          />
        </div>
        <div
          class="pinned"
          style="position: absolute; pointer-events: none; height: 9px; width: 100%; left: 0px; visibility: hidden;"
        >
          <div
            style="height: 1px; width: 100%; top: 4.5px; position: absolute;"
          />
        </div>
      </div>
      <div
        class="track"
        style="display: flex; flex: 3;"
      >
        <div
          class="track-container"
        >
          <canvas
            style="position: absolute;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
